# Progetto - Complementi di Basi di Dati A.A. 2023/2024

# Libraries Used in the Project

### pandas

**Purpose:** pandas is used for data manipulation and analysis, to read data from CSV files, perform data transformation and cleaning operations

### joblib

**Purpose:** joblib is a library used for saving and loading Python objects, including machine learning models. Here is used to save the model after training so that it can be easily loaded and used for predictions later.

### fuzzywuzzy

**Purpose:** fuzzywuzzy is a Python library that provides fuzzy string matching capabilities. It is used here to correct invalid entries in the data.

### SQLAlchemy

**Purpose:** SQLAlchemy is used to connect to a PostgreSQL database, load data from it and save prediction results back to the database

### scikit-learn (sklearn)

**Purpose:** is a machine learning library in Python that provides various tools for machine learning tasks. Here is used for data preprocessing, model training model evaluation and making predictions.

### scikit-optimize

**Purpose:** scikit-optimize is a library for sequential model-based optimization in Python. It is used to create hyperparameters for the model used in scikit-learn using a Bayesian optimization




# Code explaination

## Functions Used in the Code

- **get_NA_columns(data_description)**: Identifies columns where 'NA' is a valid value according to the provided data description and returns a list of such columns. This is done to differentiate between real NaN and  Na values that pandas automatically considers null.

- **validate_and_correct_value(column, value, data_description)**: Validates and corrects the value if it’s not valid for the specific column. It also applies a fuzzySearch if it thinks the value could be a typo.

- **clean_and_validate_data(dataFrame, data_description, columns_to_group)**: Iterates through every column of the DataFrame to check if the values are all correct, if it finds anything wrong the data is sent to `validate_and_correct_value`.

- **save_model(best_model, model_file)**: Saves the best-trained model to a file using `joblib`.

- **load_model(model_file)**: Loads a trained model from a file using `joblib`.

- **load_data(engine, table_name, data_description)**: Loads data from a specified table in the PostgreSQL database and replaces NaN values in columns where 'NA' is a valid value according to the data description.

- **prepare_data_for_prediction(data)**: Transforms categorical string variables into numerical values using either a `LabelEncoder` or the `ordinal_encodings`, preparing the data for machine learning models.

- **find_best_params(data)**: Finds the best parameters for the predictive model using Bayesian optimization.

- **train_model(data, model_file)**: Trains a predictive model using the generated hyperparameters from the `find_best_params` function.

- **predict_results(model, data)**: Uses the trained model to make predictions on the test data and returns a DataFrame containing the predictions.

# Dataset

The dataset used for this project is the Ames Housing Dataset, sourced from the Kaggle competition "House Prices: Advanced Regression Techniques". This dataset offers a comprehensive collection of data on residential property sales in Ames, Iowa in USA, and is widely utilized in the data science community due to its rich and diverse features, making it ideal for various types of data analysis and modeling projects.

The dataset is divided into two main files: ‘train.csv’ and ‘test.csv’. The training set includes 1460 observations with 81 features each and the test set contains 1459 observations with 80 features.

The features in this dataset span a broad range of categories, each contributing uniquely to the property's overall value. These categories include property descriptors, location features, house features, rooms and facilities, external features and miscellaneous features. Each feature category provides a different dimension of information about the properties, contributing to a holistic view necessary for accurate price prediction.

## Why did we choose this dataset?

Firstly, its richness and diversity make it ideal for a predictive modeling project. The dataset includes a comprehensive set of features that cover various aspects of a residential property, providing ample opportunities for extensive exploration and feature engineering. The dataset is well-documented and popular within the data science community. This popularity means there is a wealth of resources, discussions and prior work available, which can be helpful for reference and benchmarking. It provides a solid foundation for learning and applying advanced regression techniques. It is well-structured and of a manageable size, providing enough data to train robust models without requiring extensive computational resources.



In [1]:
import pandas as pd
from sqlalchemy import create_engine
from sqlalchemy.engine import URL
from sqlalchemy import text

# Connection parameters for the PostgreSQL server
db_url = URL.create(
    drivername="postgresql+psycopg2",
    username="postgres",
    password="unimib",
    host="localhost",
    port="5432",
    database=f"postgres"
)

# Database name
new_db_name = 'test_db'

try:
    # Creation of the engine for the connection to the database
    engine = create_engine(db_url, isolation_level="AUTOCOMMIT")
    with engine.connect() as conn:
        # Check if the database exists
        result = conn.execute(text(f"SELECT 1 FROM pg_database WHERE datname = '{new_db_name}'"))
        exists = result.scalar() is not None

        # If it does eliminate it
        if exists:
            conn.execute(text(f"DROP DATABASE {new_db_name} WITH (FORCE)"))
            print(f"Database '{new_db_name}' deleted successfully.")

        # Create the database
        conn.execute(text(f"CREATE DATABASE {new_db_name}"))
        print(f"Database '{new_db_name}' successfully created.")

except Exception as e:
    print(f"Errore: {e}")

Database 'test_db' deleted successfully.
Database 'test_db' successfully created.


# CRUD analysis
CRUD stands for create, read, update, and delete. These are the four basic operations that can be performed on data in a database:

- **create** - inserting new records into the database
- **read** - retrieving existing records from the database
- **update** - modifying existing records in the database
- **delete** - removing records from the database

This analysis involves examining these operations to understand how data flows through the system and how it is manipulated. This analysis helps in designing, optimizing, and maintaining database systems.

**In our project:**

- **create:** During the initial phase of the project, we used the `create_engine` function from SQLAlchemy to establish a connection to the PostgreSQL database and then inserted the cleaned and prepared data into the database.

- **read:** Throughout the project, we frequently retrieved data from the PostgreSQL database for analysis and modeling. We used Pandas' `read_sql` function to execute SQL queries and load data into data frames.

- **update:** We implemented data validation and correction functions to ensure data integrity. When invalid or missing values were detected, they were corrected or replaced.

### Connect to PostgreSQL
- **Connect to PostgreSQL**: Establishes a connection to a PostgreSQL database named `test_db` running locally with username `postgres` and password `unimib`.

In [2]:
engine = create_engine('postgresql://postgres:unimib@localhost/test_db')

## Create Table
- **Create Table**: Creates a DataFrame `dataFrame` with columns `Id`, `A`, `B`, and `C`, and saves it to a table named `test_table` in the PostgreSQL database. If the table already exists, it will be replaced.

In [3]:
data = {
    'Id': [1, 2, 3],
    'A': ['foo', 'bar', 'baz'],
    'B': [True, False, True],
    'C': [float('NaN'), float('NaN'), float('NaN')]
}

dataFrame = pd.DataFrame(data)
table_name = "test_table"
dataFrame.to_sql(table_name, engine, index=False, if_exists='replace')
print(f"Table '{table_name}' successfully created.")

Table 'test_table' successfully created.


### Read from Table

- **Read from Table**: Executes SQL queries to fetch all rows from `test_table` and prints the resulting DataFrame, and another query that fetches rows where `Id` is equal to 1.

In [4]:
table_name = "test_table"
query = f'SELECT * FROM {table_name}'
dataFrame = pd.read_sql(query, engine)
print(dataFrame)

   Id    A      B     C
0   1  foo   True  None
1   2  bar  False  None
2   3  baz   True  None


In [5]:
table_name = "test_table"
query = f'SELECT * FROM {table_name} WHERE "Id" = 1'
dataFrame = pd.read_sql(query, engine)
print(dataFrame)

   Id    A     B     C
0   1  foo  True  None


## Update Table
- **Update Table**: Updates the `C` column in `test_table` based on values in the `data` dictionary using SQL `UPDATE` statements. Uses SQLAlchemy's `text` construct to write SQL queries with parameters and executes them through the PostgreSQL connection.

In [6]:
table_name = "test_table"
data = {
    'Id': [1, 2, 3],
    'C': [1.2, 3.4, 8.0]
}

dataFrame = pd.DataFrame(data)

with engine.connect() as connection:
    for index, row in dataFrame.iterrows():
        update_query = text(f"""
            UPDATE {table_name}
            SET "C" = :c
            WHERE "Id" = :id
        """)
        
        params = {
            'id': row['Id'],
            'c': row['C']
        }
        
        connection.execute(update_query, params)

    connection.commit()

table_name = "test_table"
query = f'SELECT * FROM {table_name}'
dataFrame = pd.read_sql(query, engine)
print(dataFrame)

   Id    A      B    C
0   1  foo   True  1.2
1   2  bar  False  3.4
2   3  baz   True  8.0


## Delete Table
- **Delete Table**: Updates the `C` column in `test_table` based on values in the `data` dictionary using SQL `UPDATE` statements. Uses SQLAlchemy's `text` construct to write SQL queries with parameters and executes them through the PostgreSQL connection.

In [7]:
table_name = "test_table"

with engine.connect() as connection:
    drop_query = text(f'DROP TABLE IF EXISTS {table_name}')
    connection.execute(drop_query)
    connection.commit()
    print(f"Table '{new_db_name}' deleted successfully.")


Table 'test_db' deleted successfully.


# Database Systems Comparison

### PostgreSQL

PostgreSQL is a powerful and open-source relational database system known for its reliability and stability. It excels in handling complex queries and supports a wide range of data types. We chose PostgreSQL for our project due to its robustness, which is essential when dealing with large datasets. It seamlessly integrates with Python, facilitating smooth data loading and management. Overall, PostgreSQL is a solid choice for storing and analyzing our housing data.

### Other DBMS

- **MongoDB:** MongoDB is a NoSQL, document-oriented database renowned for its flexibility and scalability. Unlike traditional relational databases, MongoDB stores data in JSON-like documents, allowing for dynamic schema design and easy handling of unstructured or semi-structured data. It is highly suitable for projects requiring rapid development and frequent data format changes. MongoDB's horizontal scalability enables it to handle large volumes of data and high demand scenarios effectively.

- **Neo4j:** Neo4j is a graph database management system designed for storing and querying data in graph structures. It utilizes nodes, relationships, and properties to represent and store data, making it ideal for applications with complex and interconnected data, such as social networks and recommendation engines. Neo4j's optimized graph storage and processing engine provide efficient querying and traversal capabilities that are challenging to achieve with traditional relational databases.

- **MySQL:** MySQL is an open-source relational database management system widely used for web applications and data warehousing. It follows the traditional relational model, organizing data into tables with rows and columns. MySQL supports standard SQL for querying and managing data, making it popular for applications needing to manage complex transactions and ensure data consistency. Known for its reliability, ease of use, and strong community support, MySQL offers various storage engines to cater to diverse performance and feature requirements.


| Feature          | PostgreSQL                                                   | MongoDB                                                       | Neo4j                                                         | MySQL                                                         |
|------------------|--------------------------------------------------------------|---------------------------------------------------------------|---------------------------------------------------------------|---------------------------------------------------------------|
| Data Model       | Relational (tables)                                          | Document-oriented (JSON-like documents)                        | Graphs (nodes, relationships)                                  | Relational (tables)                                           |
| Query Language   | SQL                                                          | JSON-based Query Language                                      | Cypher                                                        | SQL                                                           |
| Schema           | Fixed schema, requires migration for changes                 | Flexible schema, dynamic and can evolve over time              | Schema-less for nodes and relationships                        | Fixed schema, requires migration for changes                  |
| Scalability      | Vertical and horizontal scaling                              | Horizontal scaling                                             | Horizontal scaling                                             | Vertical and horizontal scaling                               |
| Use Cases        | Financial systems, enterprise applications, complex queries  | Content management, IoT applications, high write loads         | Social networks, recommendation engines, complex relationships | Web applications, content management systems, reliable transactional systems |
| Data Integrity   | High                                                         | Moderate                                                      | Moderate                                                      | High                                                          |
| Performance      | High performance, especially for read-heavy operations       | High performance for write-heavy and unstructured data workloads | Optimized for graph operations and traversals                  | High performance, especially for read-heavy operations        |
| ACID Compliance  | Yes                                                          | Yes (with limitations)                                         | Yes                                                           | Yes                                                           |
| Strengths        | Strong consistency and reliability, rich SQL features        | Flexible schema design, high scalability, good for hierarchical data | Intuitive graph traversal, efficient for connected data       | Wide adoption, strong community support, good for reliable transaction processing |
| Weaknesses       | Can be complex to scale horizontally, may be slower with very large datasets | Less efficient for complex queries compared to SQL databases   | Less mature ecosystem compared to relational DBs, complexity in scaling | Limited support for complex data types, less performant for large-scale write operations |

To summarize, PostgreSQL was chosen for this project for several compelling reasons:

- **Integration with Pandas:** PostgreSQL seamlessly integrates with the Pandas library, facilitating efficient data manipulation and analysis tasks throughout the project.

- **Data Integrity and Reliability:** PostgreSQL ensures robust data integrity with ACID compliance, which is crucial for maintaining transactional consistency and ensuring accurate forecasting of housing prices.

- **Flexibility:** PostgreSQL offers extensive support for various data types and advanced features, providing the flexibility to handle diverse datasets and complex data structures effectively.

- **Extensibility:** PostgreSQL's extensible nature allows for the integration of custom functions, data types, and operators, enabling tailored solutions to specific project requirements.

- **Performance:** PostgreSQL is recognized for its high performance, particularly in read-heavy operations. This capability is advantageous given the project's large dataset and complex query requirements.

- **Strong Community and Documentation:** PostgreSQL benefits from a vibrant community and comprehensive documentation, ensuring ample resources for support, best practices, and continuous learning throughout the project's lifecycle.


# Query and Timing

Here are some examples of different queries using our dataset

In [8]:
import time
# Connect to postgreSQL
engine = create_engine('postgresql://postgres:unimib@localhost/ames_housing_db')

test_table = "housing_data_test"
train_table = "housing_data_train"
results_table = "housing_data_prediction"

0. **Query printing the whole database:**
    - A normal select query without any condition asking for the entire table (`train_table`).

In [9]:
start_time = time.time()
query = f'SELECT * FROM {train_table}'
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

        Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0        1          60       RL         65.0     8450   Pave    NA      Reg   
1        2          20       RL         80.0     9600   Pave    NA      Reg   
2        3          60       RL         68.0    11250   Pave    NA      IR1   
3        4          70       RL         60.0     9550   Pave    NA      IR1   
4        5          60       RL         84.0    14260   Pave    NA      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1455  1456          60       RL         62.0     7917   Pave    NA      Reg   
1456  1457          20       RL         85.0    13175   Pave    NA      Reg   
1457  1458          70       RL         66.0     9042   Pave    NA      Reg   
1458  1459          20       RL         68.0     9717   Pave    NA      Reg   
1459  1460          20       RL         75.0     9937   Pave   nan      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fe

1. **Query aggregating data:**
   - Calculate the average sale price (`SalePrice`) for each street condition (`Street`).


In [10]:
start_time = time.time()
query = f'''
   SELECT "Street", AVG("SalePrice") AS AvgSalePrice
   FROM {train_table}
   GROUP BY "Street";
'''
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

  Street   avgsaleprice
0   Pave  181130.538514
1   Grvl  130190.500000
Time elapsed: 0.0020s


2. **Query with nested query:**
   - Find the house with the highest sale price (`SalePrice`) and its neighborhood (`Neighborhood`).


In [11]:
start_time = time.time()
query = f'''
   SELECT "Neighborhood", "SalePrice"
   FROM {train_table}
   WHERE "SalePrice" = (SELECT MAX("SalePrice") FROM {train_table});
'''
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

  Neighborhood  SalePrice
0      NoRidge   755000.0
Time elapsed: 0.0024s


3. **Query with multiple operations:**
   - Compute the absolute difference and percentage error between predicted sale price (`SalePrice_predicted`) and actual sale price (`SalePrice_actual`) for the test dataset.

In [12]:
start_time = time.time()
query = f'''
    SELECT
        "Id",
        "SalePrice_actual",
        "SalePrice_predicted",
        ABS("SalePrice_actual" - "SalePrice_predicted") AS AbsError,
        ABS(("SalePrice_actual" - "SalePrice_predicted") / "SalePrice_actual") * 100 AS PercentError
    FROM {results_table};
'''
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

        Id  SalePrice_actual SalePrice_predicted abserror percenterror
0     1461     169277.052498                None     None         None
1     1462     187758.393989                None     None         None
2     1463     183583.683570                None     None         None
3     1464     179317.477511                None     None         None
4     1465     150730.079977                None     None         None
...    ...               ...                 ...      ...          ...
1454  2915     167081.220949                None     None         None
1455  2916     164788.778231                None     None         None
1456  2917     219222.423400                None     None         None
1457  2918     184924.279659                None     None         None
1458  2919     187741.866657                None     None         None

[1459 rows x 5 columns]
Time elapsed: 0.0035s


4. **Complex query with JOIN and aggregation:**
   - Calculate the average absolute error (`AvgAbsError`) and average percentage error (`AvgPercentError`) for each neighborhood (`Neighborhood`) by joining the training dataset with the results dataset on matching IDs (`Id`).

In [13]:
start_time = time.time()
query = f'''
    SELECT train."Neighborhood",
        AVG(result."AbsError") AS AvgAbsError,
        AVG(result."PercentError") AS AvgPercentError
    FROM {train_table} train
    JOIN {results_table} result ON train."Id" = result."Id"
    GROUP BY train."Neighborhood"
    ORDER BY AvgPercentError DESC;
'''
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

Empty DataFrame
Columns: [Neighborhood, avgabserror, avgpercenterror]
Index: []
Time elapsed: 0.0020s


5. **Query using aggregate functions and filters:**
   - Count the number of houses with a pool (`PoolArea > 0`) for each zoning type (`MSZoning`) in the test dataset.

In [14]:
start_time = time.time()
query = f'''
    SELECT "MSZoning", COUNT(*) AS NumHousesWithPool
    FROM {test_table}
    WHERE "PoolArea" > 0
    GROUP BY "MSZoning";
'''
dataFrame = pd.read_sql(query, engine)
time_elapsed = time.time() - start_time
print(dataFrame)
print(f"Time elapsed: {time_elapsed:.4f}s")

  MSZoning  numhouseswithpool
0       RL                  6
Time elapsed: 0.0030s


When working with databases, especially with smaller dimensions like those containing just over 1500 records per table, the performance of queries can vary significantly depending on how data is retrieved and processed.

In this scenario, simple SELECT queries execute quickly, typically completing in milliseconds (e.g., from 0.0035 seconds to 0.0300 seconds). However, as the size of data or complexity of operations increases, the time taken can become more substantial.



### Contributors

**Roberto** **Boccaccio** **869135**.

**Weronika** **Pyrka** **908115**.
